svgnative/src/ports/gdiplus/GDIPlusSVGRenderer.cpp

```diff
@@ -223,21 +223,27 @@ void GDIPlusSVGRenderer::Save(const GraphicStyle& graphicStyle)
 
     if (graphicStyle.clippingPath)
     {
-        // TODO
-        SVG_ASSERT_MSG(false, "Need to implement svg clipping path!");
+        std::unique_ptr<Gdiplus::GraphicsPath> clip_path(dynamic_cast<const GDIPlusSVGPath*>(graphicStyle.clippingPath->path.get())->GetGraphicsPath().Clone());
+        if (graphicStyle.clippingPath->transform)
+        {
+            const Gdiplus::Matrix* matrix = &dynamic_cast<GDIPlusSVGTransform*>(graphicStyle.clippingPath->transform.get())->GetMatrix();
+            clip_path->Transform(matrix);
+        }
+        clip_path->SetFillMode(graphicStyle.clippingPath->clipRule == WindingRule::kNonZero ? Gdiplus::FillMode::FillModeWinding : Gdiplus::FillMode::FillModeAlternate);
+        mContext->SetClip(clip_path.get(), Gdiplus::CombineModeIntersect);
     }
 
-    if (graphicStyle.opacity < 1.0f)
-    {
-        // TODO
-        SVG_ASSERT_MSG(false, "Need to implement svg opacity!");
-    }
+    float opacity = 1.0;
+    if (mOpacityStack.size() > 0)
+        opacity = mOpacityStack.top();
+    mOpacityStack.push(opacity * graphicStyle.opacity);
```

Comment:
@dirkschulze If this seems the right way to do it, I can replicate the same in Cairo. Rendering results seem accurate to me.

Reply:
@moazin Cairo graphics uses a state machine and has a save/restore logic. For simple opacity/clipping there is no additional logic needed. No need to multiply with ancestor opacities.